arch/sim/src/sim/up_linuxi2cbus.c

```diff
@@ -0,0 +1,332 @@
+/****************************************************************************
+ * arch/sim/src/sim/up_linuxi2cbus.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/types.h>
+#include <sys/ioctl.h>
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "up_linuxi2cbus.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ERROR(fmt, ...) \
+        syslog(LOG_ERR, "up_linuxi2c_host: " fmt "\n", ##__VA_ARGS__)
+#define INFO(fmt, ...) \
+        syslog(LOG_ERR, "up_linuxi2c_host: " fmt "\n", ##__VA_ARGS__)
+#define DEBUG(fmt, ...)
+
+/* Define the Linux IOCTL interfaces again here so we do not require having
+ * the kernel headers to build this.  This interface should be fairly
+ * stable so it should not cause any issues.  The naming is preserved in case
+ * we want to change it in the future.
+ */
+
+/* Linux msg flags */
+
+#define I2C_M_RD               0x0001
+#define I2C_M_TEN              0x0010
+#define I2C_M_DMA_SAFE         0x0200
+#define I2C_M_RECV_LEN         0x0400
+#define I2C_M_NO_RD_ACK        0x0800
+#define I2C_M_IGNORE_NAK       0x1000
+#define I2C_M_REV_DIR_ADDR     0x2000
+#define I2C_M_NOSTART          0x4000
+#define I2C_M_STOP             0x8000
+
+/* Linux ioctl */
+
+#define I2C_RETRIES            0x0701
+#define I2C_TIMEOUT            0x0702
+#define I2C_SLAVE              0x0703
+#define I2C_SLAVE_FORCE        0x0706
+#define I2C_TENBIT             0x0704
+#define I2C_FUNCS              0x0705
+#define I2C_RDWR               0x0707
+#define I2C_PEC                0x0708
+#define I2C_SMBUS              0x0720
+
+/* NuttX msg flags (see i2c_master.h) */
```

Comment:
it's better to create a header file and copy the required info from nuttx/i2c/i2c_master.h to it, then other host can reuse the definition.

Reply:
I really don't want to pull that header in here, both because of the additional build requirements and side effects of the header.  These are quite stable.

Reply:
But the previous comment was to move them out of the header and into the C file.

Reply:
Sorry, I mean Linux specific macro or function, not NuttX specific. NuttX definition could be shared by different host implementation, a common header file coud avoid the duplication. On the other hand, Linux definition is only used by up_linuxi2cbus.c, so it's better to put in source file directly.

Reply:
I'm fine doing that but I need to change the naming of the NuttX or Linux flags since there is conflict. If I put the NuttX ones in a common place I feel like we should still prefix them so that we don't have conflicts against host naming. Does that make sense?

Reply:
Yes, that's one of reason we can't directly use nuttx/i2c/i2c_master.h, here is the same approach for hostfs:
https://github.com/apache/incubator-nuttx/blob/master/include/nuttx/fs/hostfs.h

Reply:
Yeah I saw that. I don't really like that we leaked in the sim specific logic in that driver, but I also get that that driver is expected to work in the sim so I guess it is fine.

Reply:
Done.
There is now up_i2cbus.h that contains the NuttX structs and flags that can be used by multiple host implementations.  The Linux interface stuff all comes in via included kernel headers.